### PR TITLE
Generate the Travis config for more control over the matrix

### PR DESCRIPTION
**Pull request by matthewd**
This revisits something I previously considered in #29052. There, the primary goal was to try out Stages, and I ended up programmatically generating the Travis config.

This seems like a milder option: we still configure it with a very familiar-looking yaml file -- it just allows some more env-specific settings.

To get a good feel for how it'd really look, I've gone ahead and split up the existing settings where it seemed to make sense... hopefully I've guessed most of them correctly.

Related to #31339.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,132 +1,205 @@
...
-    - "GEM=guides"
-    - "GEM=ac:integration"
-
-rvm:
-  - 2.2.8
-  - 2.3.5
-  - 2.4.2
-  - ruby-head
-
+  - GEM=see-rvm-matrix
```
Comment (rafaelfranca) [inline .travis.yml]: Should this be here?

**Comment by sgrif** [comment]
Can we configure rubocop or some other linter to fail if either the source or generated file are edited without the other also being edited?

**Comment by matthewd** [comment]
> Can we configure rubocop or some other linter to fail if either the source or generated file are edited without the other also being edited?

Good point. How's this? https://travis-ci.org/rails/rails/jobs/315926111#L751

File: .travis.yml (hunk 1 above)
Comment (matthewd) [reply .travis.yml]: Nope ✂️

**Comment by sgrif** [comment]
Seems fine to me.

**Review by rafaelfranca — APPROVED** [review]

File: .travis.source.yml [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    before_install+:
+      - "[[ -z $encrypted_8a915ebdd931_key && -z $encrypted_8a915ebdd931_iv ]] || openssl aes-256-cbc -K $encrypted_8a915ebdd931_key -iv $encrypted_8a915ebdd931_iv -in activestorage/test/service/configurations.yml.enc -out activestorage/test/service/configurations.yml -d"
+  - env: "GEM=as PRESERVE_TIMEZONES=1"
+  - env: "GEM=ar:mysql2"
+  - env: "GEM=ar:sqlite3"
+  - env: "GEM=ar:postgresql"
+  - env: "GEM=guides"
+  - env: "GEM=ac:integration"
+  - env: "GEM=aj:integration"
+    services+:
```
Comment (kaspth) [inline .travis.source.yml]: I don't understand `+`. How does it affect `env: "GEM=aj:integration"`?

Comment (matthewd) [reply .travis.source.yml]: https://github.com/rails/rails/blob/8c290251e8d98d16528a0158656e2777d7c60a00/tasks/travis.rb#L8-L10

i.e.

https://github.com/rails/rails/blob/8c290251e8d98d16528a0158656e2777d7c60a00/.travis.source.yml#L12-L14

\+

https://github.com/rails/rails/blob/8c290251e8d98d16528a0158656e2777d7c60a00/.travis.source.yml#L56-L57

=

https://github.com/rails/rails/blob/8c290251e8d98d16528a0158656e2777d7c60a00/.travis.yml#L128-L131

Comment (kaspth) [reply .travis.source.yml]: Ahhh, nice 👍

**Pull request closed without merging**